### PR TITLE
Session-scoped submissionID Bean

### PR DESCRIPTION
**Pull request by sree-cfa**
Adding session scoped bean to help with issues around session getting dropped.

---

Adding a session scoped bean to help with tracking a clients application data. This will add another reference point for the submission ID that has been getting dropped/changed mid-application.
This is a proof-of-concept change that will help us determine if session scoped beans are more reliable than the direct httpsession management

[Background on the issue](https://docs.google.com/document/d/1HoayIzrrknps6Ls6CJwrxNYQGVzlOmuUplzl6sK2ZDw/edit#heading=h.239ypekz29uu)

### CODE REVIEW

**Comment by bseeger** [comment]
Can you describe some about how this is expected to work and how we might test it? 

I went to the heroku site, went through the flow up through the `personalInformation` page and cleared my session on the `homeAddress` page. I then filled it out and pressed continue. I got this page: 

Screenshot 2024-01-09 at 10 54 55 AM

**Comment by sree-cfa** [comment]
> Can you describe some about how this is expected to work and how we might test it?
> 
> I went to the heroku site, went through the flow up through the `personalInformation` page and cleared my session on the `homeAddress` page. I then filled it out and pressed continue. I got this page:
> 
> Screenshot 2024-01-09 at 10 54 55 AM
> I'm pretty sure that's due to a CSRF check failure.

Yeah - the Scoped bean is a more spring-managed version of what we're doing with the httpSession and attributes. Im not sure how to test it since Im not sure how to recreate this issue. The idea is that the bean is managed by spring context and would be more reliable/persistent than the httpSession data. If the cache/cookies/session data is cleared, then that bean should be cleared as well

**Comment by bseeger** [comment]
Ah, okay, I see what you mean. This is on par with what we haven't and isn't really session recovery.  It's just switching out what we have for something we home is more stable.  Sorry, I was getting confused as the code tries to recover the session.

File: src/main/java/org/ladocuploader/app/interceptors/DataRequiredInterceptor.java [hunk 1]
```diff
@@ -6,79 +6,134 @@
...
-        this.submissionRepositoryService = submissionRepositoryService;
-    }
+  public static final String PATH_FORMAT = "/flow/{flow}/{screen}";
+  private final SubmissionRepositoryService submissionRepositoryService;
+  private final RecoverySubmissionIDs recoveryRecoverySubmissionIDs;
+
+  // Checks at signposts
+  private static final Map<String, String> REQUIRED_DATA = Map.of(
+      "signPost", "parish",
+      "householdSignPost", "parish",
```
Comment (bseeger) [inline src/main/java/org/ladocuploader/app/interceptors/DataRequiredInterceptor.java:29]: I think I'm a little uncomfortable with only doing data checks on certain pages. If I mistakenly jump into the flow somewhere, start entering data, then hit one of these pages, I will bet bounced back and never understand what happened.

Comment (bseeger) [reply src/main/java/org/ladocuploader/app/interceptors/DataRequiredInterceptor.java:29]: If the session split still happens the user experience here will be interesting.

**Comment by bseeger** [comment]
I don't know how to really evaluate this - the code looks okay. I think it might be worth having @spokenbird take a peek at this when he's back.

**Comment by sree-cfa** [comment]
@bseeger  Re comments - Yeah, I was trying not to be a perfectionist about it. I think PEBT has checks at signposts too, we can check on every page though. I dont see any issues with that. Either way, we should find out pretty quickly if this solves our issue or not. We talked about a couple more low-touch workarounds earlier as well

File: src/main/java/org/ladocuploader/app/config/SessionConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+@Configuration
+public class SessionConfiguration {
+
+  @Bean
+  @Scope(
+      value = SCOPE_SESSION,
+      proxyMode = TARGET_CLASS
+  )
+  public RecoverySubmissionIDs recoveryRecoverySubmissionIDs() {
```
Comment (spokenbird) [inline src/main/java/org/ladocuploader/app/config/SessionConfiguration.java:19]: Should this just be recoverySubmissionIDs for the method name?

Comment (sree-cfa) [reply src/main/java/org/ladocuploader/app/config/SessionConfiguration.java:19]: yes 🙃

**Comment by spokenbird** [comment]
This is interesting -- The code looks good but I honestly don't know enough about how Spring manages the session scoped bean to know if this will work or not without trying it 😅 

I think the ultimate test though would be to change Submission in the FFB to be a session scoped bean as in that case we would get all the benefits of Spring automatically managing the session life cycle and the submission object for us?

**Comment by sree-cfa** [comment]
@spokenbird 

Yeah, I also could not pin-point what spring would be doing to retain the session
> This is interesting -- The code looks good. I like the idea of having a session scoped bean that's managed by spring to recover dropped ID's but I honestly don't know enough about how Spring manages the session scoped bean to know if this will work or not without trying it 😅

Right - this PR is a low-lift proof of concept. If the bean ends up being more reliable, we can make the change to FFL 🤞 
> I think the ultimate test though would be to change Submission in the FFB to be a session scoped bean as in that case we would get all the benefits of Spring automatically managing the session life cycle and the submission object for us?

**Pull request merged**